### PR TITLE
Fix feature gating

### PR DESCRIPTION
**Pull request by tcharding**
Currently we have a few problems with our feature gating, attempt to audit all feature gating and fix it by doing:

1. Do not enable features on optional dependencies (`rand` and `bitcoin-hashes`) in dev-dependencies, doing so hides broken feature gating in unit tests.
2. Do not use unnecessary feature combinations when one feature enables another e.g. `any(feature = "std", feature = "alloc")`.
3. Enable "std" from "rand-std" and "bitcoin-std" (and fix features gating as for point 2).
4. Clean up code around `rand::thread_rng`, this is part of this patch because `thread_rng` requires the "rand-std" feature.
5. Clean up CI test script to test each feature individually now that "rand-std" and "bitcoin-hashes-std" enable "std".

### CODE REVIEW

**Review by apoelstra — APPROVED** [review]
ACK 9c748550b4435fc4b6c5f6618f4deb2f2553f118

**Comment by apoelstra** [comment]
Will wait on @Kixunil to sanity-check before merging. This is a big PR but it's almost entirely mechanical and all the "real" changes are in `Cargo.toml`.

**Comment by tcharding** [comment]
CI seems to be running for a long time, maybe I broke something?

**Comment by tcharding** [comment]
> Will wait on @Kixunil to sanity-check before merging. This is a big PR but it's almost entirely mechanical and all the "real" changes are in `Cargo.toml`.

And I could easily have missed some places, I was relying on grep heavily to find lines of code that needed patching.

File: src/lib.rs
Comment (Kixunil) [inline src/lib.rs:541]: IIUC these do not require `alloc` but only tests requiring `alloc` happen to use them, right?

It's not very readable but the only somewhat reasonable idea beyond adding a comment that comes to my mind is having `test_std`, `test_alloc`, `test_rand`... etc modules and the modules have appropriate cfg. Maybe not having same cfgs multiple times is good enough benefit for it?

File: src/scalar.rs
Comment (Kixunil) [inline src/scalar.rs:47]: Oh, I should've implemented `Distribution<Scalar> for Standard`. :man_facepalming:

**Review by Kixunil — APPROVED** [review]
ACK 9c748550b4435fc4b6c5f6618f4deb2f2553f118

Very nice cleanup!

All of my suggestions are intended for possible future PRs.

File: src/ecdh.rs
Comment (apoelstra) [reply src/ecdh.rs:36]: Agreed. The old formatting was chosen with "this line will be hidden" in mind.

I have a mild preference for indenting vs putting `{` on the next line.

File: src/lib.rs
Comment (apoelstra) [reply src/lib.rs:528]: This function is only available within the crate, where we routinely need 32 uniformly random bytes and never need any other quantity of differently-random bytes.

File: src/lib.rs
Comment (apoelstra) [reply src/lib.rs:541]: I think this is a pretty common pattern in test modules to have feature-gated imports because there are feature-gated tests.

File: src/scalar.rs
Comment (apoelstra) [reply src/scalar.rs:47]: ITYM for `Uniform`, but honestly this seems like more complexity (and reading `rand` docs) (and hoping they never change) than it's worth.

Comment (Kixunil) [reply src/scalar.rs:47]: I think for both `Standard` and `Uniform`. If `rand` is still too far away from stabilizing maybe we should just put it off until then.

**Comment by apoelstra** [comment]
Thanks for your review and (somewhat hesitant) ACK! We'll address things in followup PRs.

File: src/ecdh.rs
Comment (tcharding) [reply src/ecdh.rs:36]: Its already laborious to write example code in comments because of zero tooling support, we should be careful what we choose because we (I) will have to go back and change every example in the whole codebase.

**Comment by tcharding** [comment]
Thanks fellas, feels like the feature gating saga is going to be an ongoing one :)

File: src/ecdh.rs
Comment (apoelstra) [reply src/ecdh.rs:36]: Ok, then let's do `{` on the next line since that's easiest :P and say that this is a very low priority thing.

**Comment by Kixunil** [comment]
Should we release a new version so that we unblock development in `bitcoin`?

**Comment by apoelstra** [comment]
I'd like to get #518 in as well since I believe it is also semver-breaking.

**Comment by Kixunil** [comment]
Oh, yes, we had "remove deprecated" commits. Would it make sense to backport this for point release?

**Comment by apoelstra** [comment]
I **think** so. Normally changes to feature flags are breaking but I think this is purely additive (except for dev-dependencies). Similarly I think all the changes to feature-gating of functionality is purely additive (things are now available that didn't use to be).

Because it touches so many LOC it'd be a real pain to backport, I think. In particular #499 will be really annoying to backport past :).

I think it'd be less work to get #518 in and then cut a release -- it's almost ready.

**Comment by Kixunil** [comment]
Oh, OK. Yes, this change is backwards-compatible.

**Comment by apoelstra** [comment]
Ok, 518 is in. We can just do a major release and pull that into rust-bitcoin I think.

**Comment by apoelstra** [comment]
cc @tcharding can you open a release PR so that I can ack and merge it?

**Comment by tcharding** [comment]
Sure thing!

**Pull request merged**
